### PR TITLE
[StatsMigration] Add mysql tables for partition class reports

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
We already migrated account stats report from helix to mysql, this PR would start the first step to migrate partition class report from helix to mysql by adding mysql tables.
This is the design doc for partition class report migration https://docs.google.com/document/d/1obzRunMrcA0BuBMLw42tw87gjzlGeZW4gpSw-4l47wI/edit#heading=h.1olkullkqvdq

A partition class report include account/container storage usage and keep them under different partition class names. A typical partition class report looks like this
```
{
    “max-replicas-all-datacenter”:{
        "Partition[11548]":{
            "A[1085]___C[8]":{"v":531308584},
            "A[1174]___C[8]":{"v":1443934},
            "A[1174]___C[9]":{"v":638918}
       }
    },
    “Video-31”:{
        "Partition[11657]":{
            "A[1085]___C[8]":{"v":524193383},
            "A[1174]___C[9]":{"v":1781907},
            "A[1186]___C[9]":{"v":153613},
            "A[154]___C[3]":{"v":12735027},
            "A[320]___C[8]":{"v":18379013},
            "A[320]___C[9]":{"v":1599623}
        }
    }
}
```
As you can see, the account/container usage data is exactly the same data as the AccountReoprt, so we can keep those storage usage data in AccountReports table. After that, we need to save partition and partition class metadata in other tables.

Partition class names like "max-replicas-all-datacenter", "video-31" would be saved in PartitionClassNames table. Partition ids would be saved in Partitions table. So to fetch the partition report for a given host:
```
1. select * from AccountReports where clusterName = clusterName and hostname = host; -> return all containers's usages
2. select Partitions.id, PartitionClassNames.name from Partitions INNER JOIN PartitionClassNames where Partition.partitionClassId = PartitionClassNames.id and Partitions.clusterName = PartitionClassNames.clusterName -> get all the partitions and their partition class names;
3. Separate partition ids based partition class names, and separate containers' usages based on what partition id container belongs to.
```

### CODE REVIEW

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl [hunk 1]
```diff
@@ -64,3 +64,52 @@ CREATE TABLE IF NOT EXISTS AggregatedAccountReportsMonth
...
+/**
+ * This table is created to store all the partition ids and its corresponding partition class id.
+ */
+CREATE TABLE IF NOT EXISTS Partitions (
+    clusterName VARCHAR(25) NOT NULL,
+    id INT NOT NULL,
+    partitionClassId SMALLINT NOT NULL,
+
+    PRIMARY KEY(clusterName, id),
+    FOREIGN KEY (clusterName, partitionClassId) REFERENCES PartitionClassNames(clusterName, id)
```
Comment (lightningrob) [inline ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl:97]: Does hosted mysql allow foreign keys?  I thought not.

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl [hunk 2]
```diff
@@ -64,3 +64,52 @@ CREATE TABLE IF NOT EXISTS AggregatedAccountReportsMonth
...
+ * -------------------------------------------------
+ * Ambry-Prod   |1   |max-replicas-all-datacenters |
+ * Ambry-Video  |1   |video-311                    |
+ * -------------------------------------------------
+ */
+CREATE TABLE IF NOT EXISTS PartitionClassNames
+(
+    clusterName VARCHAR(25) NOT NULL,
+    id SMALLINT NOT NULL,
+    name VARCHAR(45) NOT NULL,
```
Comment (lightningrob) [inline ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl:82]: I would call this partitionClassName.
Also, does this support multiple partition classes per cluster (which we may have in the future)?

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl [hunk 3]
```diff
@@ -64,3 +64,52 @@ CREATE TABLE IF NOT EXISTS AggregatedAccountReportsMonth
...
+
+    PRIMARY KEY(clusterName, id)
+)
+CHARACTER SET utf8 COLLATE utf8_bin;
+
+/**
+ * This table is created to store all the partition ids and its corresponding partition class id.
+ */
+CREATE TABLE IF NOT EXISTS Partitions (
```
Comment (cgtz) [inline ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl:91]: How is the Partitions table used?

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl [hunk 4]
```diff
@@ -64,3 +64,52 @@ CREATE TABLE IF NOT EXISTS AggregatedAccountReportsMonth
     month VARCHAR(25) NOT NULL
 )
 CHARACTER SET utf8 COLLATE utf8_bin;
+
+/**
+ * This table is created to keep a list of partition class names. In main cluster, we have max-replicas-all-datacenters.
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl]: updated

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl:83]: I think `name` is fine. It's table `PartitionClassNames` so we already know it should be partition classes' names.

File: ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl]: do we have a wiki about what feature it supports and what not?

Comment (cgtz) [reply ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl]: https://iwww.corp.linkedin.com/wiki/cf/display/IST/MYSQL+Eng+Support+Overview#MYSQLEngSupportOverview-WhatdoweNOTsupport

Comment (justinlin-linkedin) [reply ambry-mysql/src/main/resources/AmbryContainerStorageStats.ddl]: @cgtz thanks

**Review by lightningrob — APPROVED** [review]

**Review by cgtz — APPROVED** [review]

**Pull request merged**